### PR TITLE
Improve predict CLI for saving multitask prediction

### PR DESCRIPTION
**Pull request by hwpang**
## Description
This PR
- removes the target columns CLI input from `chemprop predict`
- makes sure the prediction.csv generated by `chemprop train` is consistent with `chemprop predict`.

## Example / Current workflow
As pointed out in https://github.com/chemprop/chemprop/issues/890, the prediction .csv generated with `chemprop train` saves the target column headers, while `chemprop predict` saves with dummy headers unless the user input the target columns via the CLI. This can be inconvenient to the user.

## Bugfix / Desired workflow
Since the purpose of .pt file is made for easy model portability and inference, I think we can also save the target columns in the .pt file. We can then use this target columns to write the prediction.csv file in `chemprop predict`. This way, the prediction.csv generated by `chemprop predict` is consistent with `chemprop train` and has the multitask column names.

## Relevant issues
https://github.com/chemprop/chemprop/issues/890

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/models/utils.py [hunk 1]
```diff
@@ -6,8 +6,15 @@
...
 
 
-def save_model(path: PathLike, model: MPNN) -> None:
-    torch.save({"hyper_parameters": model.hparams, "state_dict": model.state_dict()}, path)
+def save_model(path: PathLike, model: MPNN, target_columns: list[str] = None) -> None:
+    torch.save(
+        {
+            "hyper_parameters": model.hparams,
+            "state_dict": model.state_dict(),
+            "target_columns": target_columns,
```
Comment (davidegraff) [inline chemprop/models/utils.py:14]:
```suggestion
            "output_columns": output_columns,
```

File: chemprop/models/utils.py [hunk 2]
```diff
@@ -17,3 +24,8 @@ def load_model(path: PathLike, multicomponent: bool) -> MPNN:
         model = MPNN.load_from_file(path)
 
     return model
+
+
+def load_target_columns(path: PathLike) -> list[str] | None:
+    model_file = torch.load(path)
+    return model_file.get("target_columns", None)
```
Comment (davidegraff) [inline chemprop/models/utils.py:31]:
```suggestion
def load_output_columns(path: PathLike) -> list[str] | None:
    model_file = torch.load(path)
    
    return model_file.get("output_columns")
```

File: chemprop/cli/predict.py [hunk 3]
```diff
@@ -282,12 +277,9 @@ def make_prediction_for_models(
...
     average_preds = torch.mean(torch.stack(individual_preds).float(), dim=0)
-    if args.target_columns is not None:
-        assert (
-            len(args.target_columns) == model.n_tasks
-        ), "Number of target columns must match the number of tasks."
-        target_columns = args.target_columns
-    else:
+
+    target_columns = load_target_columns(model_paths[0])
+    if target_columns is None:
```
Comment (davidegraff) [inline chemprop/cli/predict.py:282]:
```suggestion
    output_columns = load_output_columns(model_paths[0])
    if output_columns is None:
```

Comment (davidegraff) [reply chemprop/cli/predict.py:282]: Can we just condense `load_model` and `load_target_columns` into one function? If we're saving them in the same structured file, I don't see the benefit to reading a file twice.

File: chemprop/cli/predict.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 from chemprop import data
 from chemprop.cli.common import add_common_args, process_common_args, validate_common_args
 from chemprop.cli.utils import Subcommand, build_data_from_files, make_dataset
-from chemprop.models import load_model
+from chemprop.models.utils import load_model, load_target_columns
```
Comment (davidegraff) [inline chemprop/cli/predict.py:14]:
```suggestion
from chemprop.models.utils import load_model, load_output_columns
```

File: chemprop/cli/predict.py (hunk 3 above)
Comment (hwpang) [reply chemprop/cli/predict.py:282]: Yeah, we can condense them into one function. However, it will still require reading the file twice because the model  file is read within the attribute `load_from_file`, but not in the load_model

```
def load_model(path: PathLike, multicomponent: bool) -> MPNN:
    if multicomponent:
        model = MulticomponentMPNN.load_from_file(path)
    else:
        model = MPNN.load_from_file(path)

    return model
```

File: chemprop/cli/hpopt.py [hunk 5]
```diff
@@ -462,7 +462,7 @@ def main(args: Namespace):
         smiles_cols=args.smiles_columns,
         rxn_cols=args.reaction_columns,
         target_cols=args.target_columns,
-        ignore_cols=args.ignore_columns,
+        ignore_cols=None,
```
Comment (shihchengli) [inline chemprop/cli/hpopt.py:465]: Why is this changed to None?

File: chemprop/cli/predict.py [hunk 6]
```diff
@@ -282,12 +277,9 @@ def make_prediction_for_models(
...
-    if args.target_columns is not None:
-        assert (
-            len(args.target_columns) == model.n_tasks
-        ), "Number of target columns must match the number of tasks."
-        target_columns = args.target_columns
-    else:
+
+    output_columns = load_output_columns(model_paths[0])
+    if output_columns is None:
         target_columns = [
```
Comment (shihchengli) [inline chemprop/cli/predict.py:283]:
```suggestion
        output_columns = [
```

Comment (shihchengli) [reply chemprop/cli/predict.py:283]: There are some `target_columns` that have not been replaced with `output_columns` in the file.

File: chemprop/cli/predict.py (hunk 3 above)
Comment (shihchengli) [reply chemprop/cli/predict.py]: How about always returning `output_columns` from the `load_from_file` function?

File: chemprop/cli/train.py [hunk 7]
```diff
@@ -852,6 +852,19 @@ def train_model(
...
+            args.smiles_columns,
+            args.reaction_columns,
+            args.target_columns,
+            args.ignore_columns,
+            args.splits_column,
+            args.weight_column,
+            args.no_header_row,
+        )
+        input_cols = (args.smiles_columns or []) + (args.reaction_columns or [])
+        target_cols = columns[1:] if len(input_cols) == 0 else columns[len(input_cols) :]
```
Comment (shihchengli) [inline chemprop/cli/train.py:866]: Can we overwrite the values of `target_cols` in `args` with this? This way, we don't need to pass `columns` and `target_cols` into `evaluate_and_save_predictions`.

File: chemprop/cli/predict.py (hunk 6 above)
Comment (shihchengli) [reply chemprop/cli/predict.py:283]: Shall we also change the name in other modules?

Comment (hwpang) [reply chemprop/cli/predict.py:283]: I can see if both way. For training, these are _targets_ for the model to learn, while for predicting, these are model _outputs_. So I think the current distinguished names make sense.

File: chemprop/cli/predict.py (hunk 3 above)
Comment (hwpang) [reply chemprop/cli/predict.py]: Since `load_from_file` is the attribute of the MPNN class, I think users would only expect it to have one output and returns the loaded model, but not the additional `output_columns` though. I think the current way of `load_from_file` is also nice because you can use it interchangeably with `load_from_checkpoint`.

File: chemprop/cli/train.py (hunk 7 above)
Comment (hwpang) [reply chemprop/cli/train.py:866]: Perhaps we should do this in process_common_args instead if we want to modify args

Comment (hwpang) [reply chemprop/cli/train.py:866]: hmmm also not really, even if we have `args.target_columns = target_cols`, we still need `columns` to save the predictions in `evaluate_and_save_predictions`.

Comment (shihchengli) [reply chemprop/cli/train.py]: `columns` is defined as `input_cols + target_cols`. The `input_cols` can be replaced with `input_cols = (args.smiles_columns or []) + (args.reaction_columns or [])`. So, all the information is in the args.

Comment (KnathanM) [reply chemprop/cli/train.py]: We can't do this in common.py because the file that has the column headers, `args.data_path` belongs to train. It should stay a training argument because you don't need it for predicting. 

I also recommend not overwriting `args.target_columns`, etc. because they are collected together in `format_kwargs` and passed to the data building function `build_splits` which has the same logic to get the column names. 

Instead, I see that we only need the updated `target_cols` two places, in `evaluate_and_save_predictions` and to save the model. I would generate `target_cols` (and by extension `columns`) in both places for now. It duplicates code, but we can go back in another PR and try to fix the code duplication. #882 also needs the column names, so how we get the column names could probably be improved.

File: chemprop/cli/train.py [hunk 8]
```diff
@@ -876,17 +887,17 @@ def train_model(
...
             evaluate_and_save_predictions(
-                preds, test_loader, model.metrics[:-1], model_output_dir, args
+                preds, test_loader, model.metrics[:-1], model_output_dir, args, columns
             )
 
         best_model_path = checkpointing.best_model_path
         model = model.__class__.load_from_checkpoint(best_model_path)
         p_model = model_output_dir / "best.pt"
-        save_model(p_model, model)
+        save_model(p_model, model, args.target_cols)
```
Comment (KnathanM) [inline chemprop/cli/train.py:896]:
```suggestion
        save_model(p_model, model, args.target_columns)
```
```suggestion
        save_model(p_model, model, target_cols)
```
depending on if you take my suggestion to not overwrite `args.target_columns`

File: chemprop/cli/train.py (hunk 7 above)
Comment (shihchengli) [reply chemprop/cli/train.py]: Yes, it cannot be used in common.py. I am thinking of calling `get_column_names` and overwriting `args.target_columns` right after the data building function `build_splits`. By doing so, `get_column_names` would only need to be called once, instead of being called for each `train_model` call. I am fine with generating `target_cols` in both places for now and improving it in the future.

Comment (KnathanM) [reply chemprop/cli/train.py]: I did the overwriting in process_train_args to follow your suggestion of only calling get_column_names once. I think it is safe to overwrite target columns at the very beginning because `parse_csv` in `cli/utils/parsing.py` does the same thing if target_cols is None. Now that code in parse_csv usually won't be called because we do it in process_train_args.

**Review by shihchengli — APPROVED** [review]
